packages/sdk/client/src/packlets/shell/shell-controller.ts

```diff
@@ -0,0 +1,111 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { Event } from '@dxos/async';
+import { RemoteServiceConnectionTimeout } from '@dxos/errors';
+import { PublicKey } from '@dxos/keys';
+import { schema } from '@dxos/protocols';
+import { OsShellService, ShellDisplay, ShellLayout } from '@dxos/protocols/proto/dxos/iframe';
+import { createProtoRpcPeer, ProtoRpcPeer, RpcPort } from '@dxos/rpc';
+import { createIFrame, createIFramePort } from '@dxos/rpc-tunnel';
+
+import { SHELL_CHANNEL } from './shell-runtime';
+
+const DEFAULT_SRC = 'https://halo.dxos.org/shell.html';
+
+export type ShellControllerOptions = {
+  shellSrc: string;
+  channel: string;
+  timeout: number;
+};
+
+// TODO(wittjosiah): Look at avoiding waterfalls of rpc connections.
+export class ShellController {
+  readonly spaceUpdate = new Event<PublicKey>();
+  private readonly _params;
+  private _shellRpc?: ProtoRpcPeer<{ OsShellService: OsShellService }>;
+  private _iframeId?: string;
+  private _display = ShellDisplay.NONE;
+  private _spaceKey?: PublicKey;
+
+  constructor({
+    shellSrc = DEFAULT_SRC,
+    channel = SHELL_CHANNEL,
+    timeout = 1000
+  }: Partial<ShellControllerOptions> = {}) {
+    this._params = { shellSrc, channel, timeout };
+  }
+
+  get display() {
+    return this._display;
+  }
+
+  get spaceKey() {
+    return this._spaceKey;
+  }
+
+  async setLayout(layout: ShellLayout) {
+    await this._shellRpc?.rpc.OsShellService.setLayout({ layout });
+  }
+
+  /**
+   * @internal
+   */
+  async _open(identityCreated: boolean) {
+    const port = await this._getIFramePort(identityCreated);
+
+    this._shellRpc = createProtoRpcPeer({
+      requested: {
+        OsShellService: schema.getService('dxos.iframe.OsShellService')
```

Comment:
I'd prefer to drop "OS" in the code and just use "shell".
Otherwise words like Framework or Window would be more appropriate.